pool/flexpool.go

```diff
@@ -0,0 +1,77 @@
+package pool
+
+import (
+	"sync"
+)
+
+type FlexPool struct {
+	mu      sync.Mutex
+	taskQueue map[string]chan func()
+	useCount map[string]int
+	terminating bool
+}
+
+func NewFlexPool() *FlexPool {
+	// taskQueue is unbuffered since items are always removed immediately.
+	pool := &FlexPool{
+		taskQueue:   make(map[string]chan func()),
+		useCount: make(map[string]int),
+	}
+	return pool
+}
+
+func (p *FlexPool) Terminate() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.terminating {
+		return
+	}
+	p.terminating = true
+	for _, q := range p.taskQueue {
+		close(q)
+	}
+}
+
+func (p *FlexPool) IsTerminating() bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.terminating
+}
+
+func (p *FlexPool) Submit(uid string, task func()) {
+	if p.IsTerminating() || task == nil {
+		return
+	}
+	p.mu.Lock()
+	q := p.taskQueue[uid]
+	if q == nil {
+		q = make(chan func(), 10)
+		p.taskQueue[uid] = q
+		go startWorker(q)
+	}
+	p.mu.Unlock()
+	q <- task // could block
```

Comment:
right, but also can't just move `q <- task` into the critical section, because this statement might block indefinitely, meaning that all workers are getting blocked.

I guess we also need another termination flag per task queue.